File: backend/app/api/v1/organizations.py
"""
Organization (tenant) endpoints.
"""

from time import perf_counter
from typing import Annotated, NoReturn
from uuid import UUID

import structlog
from fastapi import APIRouter, Depends, HTTPException, Query, Request, Response, status
from fastapi.responses import JSONResponse
from sqlalchemy import select

from app.api.dependencies import AsyncDB, CurrentUser, OrganizationContext, SuperAdminOnly
from app.core.user_manager import UserManager, get_user_manager
from app.main import limiter
from app.models.organization import Organization
from app.models.user import User, UserRole
from app.schemas.org_user import OrgUserCreate, OrgUserCreateRequest, OrgUserUpdate
from app.schemas.organization import (
    OrganizationArchiveRead,
    OrganizationArchiveRequest,
    OrganizationCreate,
    OrganizationPurgeForceRequest,
    OrganizationRead,
    OrganizationUpdate,
)
from app.schemas.user_fastapi import UserRead
from app.services.organization_lifecycle_service import (
    PURGE_RETENTION_DAYS,
    OrganizationLifecycleError,
    archive_organization,
    cleanup_purged_organization_storage,
    purge_force_organization,
    restore_organization,
)

router = APIRouter()
logger = structlog.get_logger(__name__)


def _raise_forbidden_superadmin_required() -> NoReturn:
    raise HTTPException(
        status_code=status.HTTP_403_FORBIDDEN,
        detail={
            "code": "FORBIDDEN_SUPERADMIN_REQUIRED",
            "message": "Superadmin required",
        },
    )


def _raise_lifecycle_error(exc: OrganizationLifecycleError) -> NoReturn:
    raise HTTPException(
        status_code=exc.status_code,
        detail={
            "code": exc.code,
            "message": exc.message,
            "details": exc.details,
        },
    ) from exc


def _raise_org_inactive_for_user_mutation(org_id: UUID) -> NoReturn:
    raise HTTPException(
        status_code=status.HTTP_409_CONFLICT,
        detail={
            "code": "ORG_NOT_ACTIVE",
            "message": "Organization is archived; cannot create or activate users",
            "details": {"org_id": str(org_id)},
        },
    )


def _ensure_superadmin_global(current_user: User) -> None:
    if not current_user.is_superuser:
        _raise_forbidden_superadmin_required()
    if current_user.organization_id is not None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={
                "code": "INVALID_ADMIN_STATE",
                "message": "Superadmin must not be scoped to an organization",
                "details": {"actor_user_id": str(current_user.id)},
            },
        )


async def _get_organization_for_update(db: AsyncDB, org_id: UUID) -> Organization | None:
    result = await db.execute(
        select(Organization).where(Organization.id == org_id).with_for_update()
    )
    return result.scalar_one_or_none()


@router.get("", response_model=list[OrganizationRead])
async def list_organizations(
    current_user: CurrentUser,
    db: AsyncDB,
    include_inactive: Annotated[bool, Query()] = False,
):
    _ensure_superadmin_global(current_user)

    query = select(Organization)
    if not include_inactive:
        query = query.where(Organization.is_active.is_(True))
    query = query.order_by(Organization.name)

    result = await db.execute(query)
    return result.scalars().all()


@router.post("", response_model=OrganizationRead, status_code=status.HTTP_201_CREATED)
async def create_organization(
    request: Request,
    data: OrganizationCreate,
    admin: SuperAdminOnly,
    db: AsyncDB,
):
    raw_payload = await request.json()
    if isinstance(raw_payload, dict) and ("is_active" in raw_payload or "isActive" in raw_payload):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={
                "code": "ORG_LIFECYCLE_FIELD_IMMUTABLE",
                "message": "New organizations are created active; use lifecycle endpoints to change state",
            },
        )

    # BaseSchema serializes with camelCase by default; ORM expects snake_case field names.
    org = Organization(**data.model_dump(by_alias=False))
    db.add(org)
    await db.commit()
    await db.refresh(org)
    return org


@router.get("/current", response_model=OrganizationRead)
async def get_current_organization(
    org: OrganizationContext,
):
    return org


@router.get("/current/users", response_model=list[UserRead])
async def list_my_org_users(
    org: OrganizationContext,
    current_user: CurrentUser,
    db: AsyncDB,
):
    """List users in my organization. Org Admin or Platform Admin only."""
    if not current_user.can_manage_org_users():
        raise HTTPException(status_code=403, detail="Not authorized to manage users")

    query = select(User).where(User.organization_id == org.id).order_by(User.email)
    result = await db.execute(query)
    return result.scalars().all()


@router.post("/current/users", response_model=UserRead, status_code=status.HTTP_201_CREATED)
async def create_user_in_my_org(
    data: OrgUserCreateRequest,
    org: OrganizationContext,
    current_user: CurrentUser,
    db: AsyncDB,
    user_manager: Annotated[UserManager, Depends(get_user_manager)],
):
    """Create user in my organization. Org Admin or Platform Admin only."""
    if not current_user.can_manage_org_users():
        raise HTTPException(status_code=403, detail="Not authorized to manage users")

    locked_org = await _get_organization_for_update(db, org.id)
    if locked_org is None:
        raise HTTPException(status_code=404, detail="Organization not found")
    if not locked_org.is_active:
        _raise_org_inactive_for_user_mutation(locked_org.id)

    # Block creating platform admins from this endpoint
    if data.role == UserRole.ADMIN.value:
        raise HTTPException(
            status_code=400,
            detail="Cannot create platform admin from this endpoint. Use /admin/users instead.",
        )

    user_create = OrgUserCreate(
        **data.model_dump(),
        organization_id=locked_org.id,
        is_superuser=False,
    )
    return await user_manager.create(user_create)


@router.get("/{org_id}", response_model=OrganizationRead)
async def get_organization(
    org_id: UUID,
    admin: SuperAdminOnly,
    db: AsyncDB,
):
    org = await db.get(Organization, org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")
    return org


@router.get("/{org_id}/users", response_model=list[UserRead])
async def list_org_users(
    org_id: UUID,
    admin: SuperAdminOnly,
    db: AsyncDB,
):
    """List users of a specific organization. Platform Admin only."""
    org = await db.get(Organization, org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")

    query = select(User).where(User.organization_id == org_id).order_by(User.email)
    result = await db.execute(query)
    return result.scalars().all()


@router.post("/{org_id}/users", response_model=UserRead, status_code=status.HTTP_201_CREATED)
async def create_org_user(
    org_id: UUID,
    data: OrgUserCreateRequest,
    admin: SuperAdminOnly,
    db: AsyncDB,
    user_manager: Annotated[UserManager, Depends(get_user_manager)],
):
    """Create user in a specific organization. Platform Admin only."""
    org = await _get_organization_for_update(db, org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")
    if not org.is_active:
        _raise_org_inactive_for_user_mutation(org.id)

    # Block creating platform admins from this endpoint
    if data.role == UserRole.ADMIN.value:
        raise HTTPException(
            status_code=400,
            detail="Cannot create platform admin from this endpoint. Use /admin/users instead.",
        )

    user_create = OrgUserCreate(
        **data.model_dump(),
        organization_id=org.id,
        is_superuser=False,
    )
    user = await user_manager.create(user_create)
    return user


@router.patch("/{org_id}", response_model=OrganizationRead)
async def update_organization(
    request: Request,
    org_id: UUID,
    data: OrganizationUpdate,
    admin: SuperAdminOnly,
    db: AsyncDB,
):
    """Update an organization. Platform Admin only."""
    raw_payload = await request.json()
    if isinstance(raw_payload, dict) and ("is_active" in raw_payload or "isActive" in raw_payload):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={
                "code": "ORG_LIFECYCLE_FIELD_IMMUTABLE",
                "message": "is_active can only be changed via lifecycle endpoints",
            },
        )

    org = await _get_organization_for_update(db, org_id)
    if not org:
        raise HTTPException(status_code=404, detail="Organization not found")

    update_data = data.model_dump(exclude_unset=True, by_alias=False)
    if update_data and not org.is_active:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={
                "code": "ORG_NOT_ACTIVE",
                "message": "Organization is archived; cannot update metadata",
                "details": {"org_id": str(org.id)},
            },
        )

    for field, value in update_data.items():
        setattr(org, field, value)

    await db.commit()
    await db.refresh(org)
    return org


@router.post("/{org_id}/archive", response_model=OrganizationArchiveRead)
@limiter.limit("10/minute")
async def archive_organization_endpoint(
    request: Request,
    org_id: UUID,
    current_user: CurrentUser,
    db: AsyncDB,
    payload: OrganizationArchiveRequest | None = None,
):
    started_at = perf_counter()
    request_id = request.headers.get("x-request-id")
    force_deactivate_users = bool(payload and payload.force_deactivate_users)

    try:
        _ensure_superadmin_global(current_user)
    except HTTPException as exc:
        logger.warning(
            "organization_archive_attempt",
            actor_user_id=str(current_user.id),
            org_id=str(org_id),
            forceDeactivateUsers=force_deactivate_users,
            deactivatedUsersCount=0,
            request_id=request_id,
            result="error",
            error_code=(
                "FORBIDDEN_SUPERADMIN_REQUIRED"
                if exc.status_code == status.HTTP_403_FORBIDDEN
                else "INVALID_ADMIN_STATE"
            ),
        )
        raise

    try:
        archive_result = await archive_organization(
            db=db,
            organization_id=org_id,
            actor_user_id=current_user.id,
            force_deactivate_users=force_deactivate_users,
        )
    except OrganizationLifecycleError as exc:
        logger.warning(
            "organization_archive_attempt",
            actor_user_id=str(current_user.id),
            org_id=str(org_id),
            forceDeactivateUsers=force_deactivate_users,
            deactivatedUsersCount=0,
            request_id=request_id,
            result="error",
            error_code=exc.code,
            error_details=exc.details,
            duration_ms=int((perf_counter() - started_at) * 1000),
        )
        _raise_lifecycle_error(exc)

    logger.info(
        "organization_archive_attempt",
        actor_user_id=str(current_user.id),
        org_id=str(archive_result.organization.id),
        org_slug=archive_result.organization.slug,
        forceDeactivateUsers=force_deactivate_users,
        deactivatedUsersCount=archive_result.deactivated_users_count,
        request_id=request_id,
        result="success",
        archived_at=(
            archive_result.organization.archived_at.isoformat()
            if archive_result.organization.archived_at
            else None
        ),
        duration_ms=int((perf_counter() - started_at) * 1000),
    )
    return {
        **OrganizationRead.model_validate(archive_result.organization).model_dump(by_alias=True),
        "deactivatedUsersCount": archive_result.deactivated_users_count,
    }


@router.post("/{org_id}/restore", response_model=OrganizationRead)
@limiter.limit("10/minute")
async def restore_organization_endpoint(
    request: Request,
    org_id: UUID,
    current_user: CurrentUser,
    db: AsyncDB,
):
    started_at = perf_counter()
    request_id = request.headers.get("x-request-id")

    try:
        _ensure_superadmin_global(current_user)
    except HTTPException as exc:
        logger.warning(
            "organization_restore_attempt",
            actor_user_id=str(current_user.id),
            org_id=str(org_id),
            request_id=request_id,
            result="error",
            error_code=(
                "FORBIDDEN_SUPERADMIN_REQUIRED"
                if exc.status_code == status.HTTP_403_FORBIDDEN
                else "INVALID_ADMIN_STATE"
            ),
        )
        raise

    try:
        organization = await restore_organization(
            db=db,
            organization_id=org_id,
        )
    except OrganizationLifecycleError as exc:
        logger.warning(
            "organization_restore_attempt",
            actor_user_id=str(current_user.id),
            org_id=str(org_id),
            request_id=request_id,
            result="error",
            error_code=exc.code,
            error_details=exc.details,
            duration_ms=int((perf_counter() - started_at) * 1000),
        )
        _raise_lifecycle_error(exc)

    logger.info(
        "organization_restore_attempt",
        actor_user_id=str(current_user.id),
        org_id=str(organization.id),
        org_slug=organization.slug,
        request_id=request_id,
        result="success",
        duration_ms=int((perf_counter() - started_at) * 1000),
    )
    return organization


@router.post("/{org_id}/purge-force", status_code=status.HTTP_204_NO_CONTENT)
@limiter.limit("5/minute")
async def purge_force_organization_endpoint(
    request: Request,
    org_id: UUID,
    payload: OrganizationPurgeForceRequest,
    current_user: CurrentUser,
    db: AsyncDB,
):
    started_at = perf_counter()
    request_id = request.headers.get("x-request-id")

    try:
        _ensure_superadmin_global(current_user)
    except HTTPException as exc:
        logger.warning(
            "organization_purge_force_attempt",
            actor_user_id=str(current_user.id),
            org_id=str(org_id),
            request_id=request_id,
            reason_present=bool(payload.reason),
            reason_length=len(payload.reason),
            ticket_id=payload.ticket_id,
            result="error",
            error_code=(
                "FORBIDDEN_SUPERADMIN_REQUIRED"
                if exc.status_code == status.HTTP_403_FORBIDDEN
                else "INVALID_ADMIN_STATE"
            ),
        )
        raise

    try:
        purge_result = await purge_force_organization(
            db=db,
            organization_id=org_id,
            confirm_name=payload.confirm_name,
            confirm_phrase=payload.confirm_phrase,
        )
    except OrganizationLifecycleError as exc:
        logger.warning(
            "organization_purge_force_attempt",
            actor_user_id=str(current_user.id),
            org_id=str(org_id),
            request_id=request_id,
            reason_present=bool(payload.reason),
            reason_length=len(payload.reason),
            ticket_id=payload.ticket_id,
            result="error",
            error_code=exc.code,
            error_details=exc.details,
            duration_ms=int((perf_counter() - started_at) * 1000),
        )
        _raise_lifecycle_error(exc)

    try:
        cleanup_status = await cleanup_purged_organization_storage(
            db=db,
            organization_id=org_id,
            cleanup_manifest_id=purge_result.cleanup_manifest_id,
            storage_paths=purge_result.storage_paths,
        )
    except OrganizationLifecycleError as exc:
        logger.warning(
            "organization_purge_force_attempt",
            actor_user_id=str(current_user.id),
            org_id=str(org_id),
            request_id=request_id,
            reason_present=bool(payload.reason),
            reason_length=len(payload.reason),
            ticket_id=payload.ticket_id,
            retention_days=PURGE_RETENTION_DAYS,
            result="error",
            error_code=exc.code,
            error_details=exc.details,
            duration_ms=int((perf_counter() - started_at) * 1000),
        )
        _raise_lifecycle_error(exc)

    logger.info(
        "organization_purge_force_attempt",
        actor_user_id=str(current_user.id),
        org_id=str(org_id),
        request_id=request_id,
        reason_present=bool(payload.reason),
        reason_length=len(payload.reason),
        ticket_id=payload.ticket_id,
        retention_days=PURGE_RETENTION_DAYS,
        archived_at=purge_result.archived_at.isoformat(),
        cleanup_status=cleanup_status,
        cleanup_manifest_id=str(purge_result.cleanup_manifest_id),
        deleted_counts=purge_result.deleted_counts,
        result="success" if cleanup_status == "completed" else "pending_cleanup",
        duration_ms=int((perf_counter() - started_at) * 1000),
    )
    if cleanup_status == "failed":
        return JSONResponse(
            status_code=status.HTTP_202_ACCEPTED,
            content={
                "error": {
                    "code": "ORG_STORAGE_CLEANUP_PENDING",
                    "message": "Organization purged from DB; storage cleanup pending manual replay",
                    "details": {
                        "org_id": str(org_id),
                        "manifest_id": str(purge_result.cleanup_manifest_id),
                    },
                }
            },
        )

    return Response(status_code=status.HTTP_204_NO_CONTENT)


# Keep "/current/..." routes before "/{org_id}/..." routes to avoid
# path shadowing (e.g. "current" matching dynamic org_id paths).
@router.patch("/current/users/{user_id}", response_model=UserRead)
async def update_my_org_user(
    user_id: UUID,
    data: OrgUserUpdate,
    org: OrganizationContext,
    current_user: CurrentUser,
    db: AsyncDB,
):
    """Update user role or status in my organization. Org Admin or Platform Admin only."""
    if not current_user.can_manage_org_users():
        raise HTTPException(status_code=403, detail="Not authorized to manage users")

    user = await db.get(User, user_id)
    if not user or user.organization_id != org.id:
        raise HTTPException(status_code=404, detail="User not found in this organization")

    locked_org = await _get_organization_for_update(db, org.id)
    if locked_org is None:
        raise HTTPException(status_code=404, detail="Organization not found")
    update_data = data.model_dump(exclude_unset=True, by_alias=False)
    if update_data and not locked_org.is_active:
        _raise_org_inactive_for_user_mutation(locked_org.id)

    # Prevent self-demotion for org admins
    if user.id == current_user.id and data.role and data.role != current_user.role:
        raise HTTPException(
            status_code=400,
            detail="Cannot change your own role.",
        )

    if user.id == current_user.id and data.is_active is False:
        raise HTTPException(
            status_code=400,
            detail="Cannot deactivate your own account.",
        )

    # Block changing to platform admin role
    if data.role == UserRole.ADMIN.value:
        raise HTTPException(
            status_code=400,
            detail="Cannot promote to platform admin from this endpoint.",
        )
    is_demoting_org_admin = (
        user.role == UserRole.ORG_ADMIN
        and update_data.get("role")
        and update_data["role"] != UserRole.ORG_ADMIN
    )
    is_deactivating_org_admin = (
        user.role == UserRole.ORG_ADMIN and update_data.get("is_active") is False
    )

    if is_demoting_org_admin or is_deactivating_org_admin:
        query = select(User.id).where(
            User.organization_id == locked_org.id,
            User.is_active,
            User.role == UserRole.ORG_ADMIN,
            User.id != user.id,
        )
        result = await db.execute(query)
        if result.first() is None:
            raise HTTPException(
                status_code=400,
                detail="Cannot remove the last active org admin.",
            )

    for field, value in update_data.items():
        setattr(user, field, value)

    await db.commit()
    await db.refresh(user)
    return user


@router.patch("/{org_id}/users/{user_id}", response_model=UserRead)
async def update_org_user(
    org_id: UUID,
    user_id: UUID,
    data: OrgUserUpdate,
    admin: SuperAdminOnly,
    db: AsyncDB,
):
    """Update user role or status in a specific organization. Platform Admin only."""
    locked_org = await _get_organization_for_update(db, org_id)
    if not locked_org:
        raise HTTPException(status_code=404, detail="Organization not found")

    user = await db.get(User, user_id)
    if not user or user.organization_id != org_id:
        raise HTTPException(status_code=404, detail="User not found in this organization")

    update_data = data.model_dump(exclude_unset=True, by_alias=False)
    if update_data and not locked_org.is_active:
        _raise_org_inactive_for_user_mutation(locked_org.id)

    # Block changing to platform admin role
    if data.role == UserRole.ADMIN.value:
        raise HTTPException(
            status_code=400,
            detail="Cannot promote to platform admin from this endpoint.",
        )
    is_demoting_org_admin = (
        user.role == UserRole.ORG_ADMIN
        and update_data.get("role")
        and update_data["role"] != UserRole.ORG_ADMIN
    )
    is_deactivating_org_admin = (
        user.role == UserRole.ORG_ADMIN and update_data.get("is_active") is False
    )

    if is_demoting_org_admin or is_deactivating_org_admin:
        query = select(User.id).where(
            User.organization_id == locked_org.id,
            User.is_active,
            User.role == UserRole.ORG_ADMIN,
            User.id != user.id,
        )
        result = await db.execute(query)
        if result.first() is None:
            raise HTTPException(
                status_code=400,
                detail="Cannot remove the last active org admin.",
            )

    for field, value in update_data.items():
        setattr(user, field, value)

    await db.commit()
    await db.refresh(user)
    return user

File: frontend/lib/api/organizations.ts
import type { User, UserRole } from "@/lib/types/user";
import { apiClient } from "./client";

export interface Organization {
	id: string;
	name: string;
	slug: string;
	contactEmail: string | null;
	contactPhone: string | null;
	isActive: boolean;
	archivedAt?: string | null;
	archivedByUserId?: string | null;
	deactivatedUsersCount?: number;
	createdAt: string;
}

export interface OrganizationCreateInput {
	name: string;
	slug: string;
	contactEmail?: string;
	contactPhone?: string;
}

export interface OrgUserCreateInput {
	email: string;
	password: string;
	firstName: string;
	lastName: string;
	role: Exclude<UserRole, "admin">;
}

export interface OrganizationUpdateInput {
	name?: string;
	contactEmail?: string | null;
	contactPhone?: string | null;
}

export interface OrganizationPurgeForceInput {
	confirmName: string;
	confirmPhrase: string;
	reason: string;
	ticketId: string;
}

export type OrganizationPurgeForceResult =
	| { status: "completed" }
	| { status: "pending_cleanup"; manifestId?: string };

export interface OrganizationArchiveInput {
	forceDeactivateUsers?: boolean;
}

export interface OrgUserUpdateInput {
	role?: Exclude<UserRole, "admin">;
	isActive?: boolean;
}

interface RawOrganizationResponse {
	id: string;
	name: string;
	slug: string;
	contact_email?: string | null;
	contactEmail?: string | null;
	contact_phone?: string | null;
	contactPhone?: string | null;
	is_active?: boolean;
	isActive?: boolean;
	archived_at?: string | null;
	archivedAt?: string | null;
	archived_by_user_id?: string | null;
	archivedByUserId?: string | null;
	deactivated_users_count?: number;
	deactivatedUsersCount?: number;
	created_at?: string;
	createdAt?: string;
}

interface RawUserResponse {
	id: string;
	email: string;
	first_name: string;
	last_name: string;
	company_name?: string;
	location?: string;
	sector?: string;
	subsector?: string;
	is_verified?: boolean;
	is_active?: boolean;
	created_at?: string;
	is_superuser?: boolean;
	role?: UserRole;
	organization_id?: string | null;
}

interface RawPurgeForcePendingResponse {
	error?: {
		code?: string;
		details?: {
			manifest_id?: string;
			manifestId?: string;
		};
	};
}

function parsePurgeManifestId(
	response: RawPurgeForcePendingResponse,
): string | undefined {
	return (
		response.error?.details?.manifestId ?? response.error?.details?.manifest_id
	);
}

function transformOrganization(
	response: RawOrganizationResponse,
): Organization {
	return {
		id: response.id,
		name: response.name,
		slug: response.slug,
		contactEmail: response.contactEmail ?? response.contact_email ?? null,
		contactPhone: response.contactPhone ?? response.contact_phone ?? null,
		isActive: response.isActive ?? response.is_active ?? true,
		archivedAt: response.archivedAt ?? response.archived_at ?? null,
		archivedByUserId:
			response.archivedByUserId ?? response.archived_by_user_id ?? null,
		deactivatedUsersCount:
			response.deactivatedUsersCount ?? response.deactivated_users_count ?? 0,
		createdAt:
			response.createdAt ?? response.created_at ?? new Date().toISOString(),
	};
}

function transformUser(response: RawUserResponse): User {
	return {
		id: response.id,
		email: response.email,
		firstName: response.first_name,
		lastName: response.last_name,
		companyName: response.company_name || undefined,
		location: response.location || undefined,
		sector: response.sector || undefined,
		subsector: response.subsector || undefined,
		isVerified: response.is_verified ?? false,
		isActive: response.is_active ?? true,
		createdAt: response.created_at || new Date().toISOString(),
		isSuperuser: response.is_superuser ?? false,
		role: response.role ?? "field_agent",
		organizationId: response.organization_id ?? null,
	};
}

export const organizationsAPI = {
	/**
	 * List all organizations (Platform Admin only)
	 */
	async list(options?: { includeInactive?: boolean }): Promise<Organization[]> {
		const searchParams = new URLSearchParams();
		if (options?.includeInactive) {
			searchParams.set("include_inactive", "true");
		}
		const query = searchParams.toString();
		const endpoint = query ? `/organizations?${query}` : "/organizations";
		const data = await apiClient.get<RawOrganizationResponse[]>(endpoint);
		return data.map(transformOrganization);
	},

	/**
	 * Get organization by ID (Platform Admin only)
	 */
	async get(orgId: string): Promise<Organization> {
		const data = await apiClient.get<RawOrganizationResponse>(
			`/organizations/${orgId}`,
		);
		return transformOrganization(data);
	},

	/**
	 * Get current organization (any authenticated user)
	 */
	async getCurrent(): Promise<Organization> {
		const data = await apiClient.get<RawOrganizationResponse>(
			"/organizations/current",
		);
		return transformOrganization(data);
	},

	/**
	 * Create a new organization (Platform Admin only)
	 */
	async create(payload: OrganizationCreateInput): Promise<Organization> {
		const body = {
			name: payload.name,
			slug: payload.slug,
			contact_email: payload.contactEmail,
			contact_phone: payload.contactPhone,
		};
		const data = await apiClient.post<RawOrganizationResponse>(
			"/organizations",
			body,
		);
		return transformOrganization(data);
	},

	/**
	 * List users of a specific organization (Platform Admin only)
	 */
	async listOrgUsers(orgId: string): Promise<User[]> {
		const data = await apiClient.get<RawUserResponse[]>(
			`/organizations/${orgId}/users`,
		);
		return data.map(transformUser);
	},

	/**
	 * Create user in a specific organization (Platform Admin only)
	 */
	async createOrgUser(
		orgId: string,
		payload: OrgUserCreateInput,
	): Promise<User> {
		const body = {
			email: payload.email,
			password: payload.password,
			first_name: payload.firstName,
			last_name: payload.lastName,
			role: payload.role,
		};
		const data = await apiClient.post<RawUserResponse>(
			`/organizations/${orgId}/users`,
			body,
		);
		return transformUser(data);
	},

	/**
	 * List users of current organization (Org Admin or Platform Admin)
	 */
	async listMyOrgUsers(): Promise<User[]> {
		const data = await apiClient.get<RawUserResponse[]>(
			"/organizations/current/users",
		);
		return data.map(transformUser);
	},

	/**
	 * Create user in current organization (Org Admin or Platform Admin)
	 */
	async createMyOrgUser(payload: OrgUserCreateInput): Promise<User> {
		const body = {
			email: payload.email,
			password: payload.password,
			first_name: payload.firstName,
			last_name: payload.lastName,
			role: payload.role,
		};
		const data = await apiClient.post<RawUserResponse>(
			"/organizations/current/users",
			body,
		);
		return transformUser(data);
	},

	/**
	 * Update an organization (Platform Admin only)
	 */
	async update(
		orgId: string,
		payload: OrganizationUpdateInput,
	): Promise<Organization> {
		const body: Record<string, unknown> = {};
		if (payload.name !== undefined) body.name = payload.name;
		if (payload.contactEmail !== undefined)
			body.contact_email = payload.contactEmail;
		if (payload.contactPhone !== undefined)
			body.contact_phone = payload.contactPhone;

		const data = await apiClient.patch<RawOrganizationResponse>(
			`/organizations/${orgId}`,
			body,
		);
		return transformOrganization(data);
	},

	async archive(
		orgId: string,
		input?: OrganizationArchiveInput,
	): Promise<Organization> {
		const body = input?.forceDeactivateUsers
			? { force_deactivate_users: true }
			: undefined;
		const data = await apiClient.post<RawOrganizationResponse>(
			`/organizations/${orgId}/archive`,
			body,
		);
		return transformOrganization(data);
	},

	async restore(orgId: string): Promise<Organization> {
		const data = await apiClient.post<RawOrganizationResponse>(
			`/organizations/${orgId}/restore`,
		);
		return transformOrganization(data);
	},

	async purgeForce(
		orgId: string,
		payload: OrganizationPurgeForceInput,
	): Promise<OrganizationPurgeForceResult> {
		const response = await apiClient.post<RawPurgeForcePendingResponse | null>(
			`/organizations/${orgId}/purge-force`,
			{
				confirm_name: payload.confirmName,
				confirm_phrase: payload.confirmPhrase,
				reason: payload.reason,
				ticket_id: payload.ticketId,
			},
		);

		if (response?.error?.code === "ORG_STORAGE_CLEANUP_PENDING") {
			const manifestId = parsePurgeManifestId(response);
			return {
				status: "pending_cleanup",
				...(manifestId ? { manifestId } : {}),
			};
		}

		return { status: "completed" };
	},

	/**
	 * Update user in a specific organization (Platform Admin only)
	 */
	async updateOrgUser(
		orgId: string,
		userId: string,
		payload: OrgUserUpdateInput,
	): Promise<User> {
		const body: Record<string, unknown> = {};
		if (payload.role !== undefined) body.role = payload.role;
		if (payload.isActive !== undefined) body.is_active = payload.isActive;

		const data = await apiClient.patch<RawUserResponse>(
			`/organizations/${orgId}/users/${userId}`,
			body,
		);
		return transformUser(data);
	},

	/**
	 * Update user in current organization (Org Admin or Platform Admin)
	 */
	async updateMyOrgUser(
		userId: string,
		payload: OrgUserUpdateInput,
	): Promise<User> {
		const body: Record<string, unknown> = {};
		if (payload.role !== undefined) body.role = payload.role;
		if (payload.isActive !== undefined) body.is_active = payload.isActive;

		const data = await apiClient.patch<RawUserResponse>(
			`/organizations/current/users/${userId}`,
			body,
		);
		return transformUser(data);
	},
};
